python/ray/data/_internal/logical/operators/from_arrow_operator.py

```diff
@@ -0,0 +1,77 @@
+from typing import TYPE_CHECKING, List, Optional, Union
+
+import ray
+from ray.data._internal.logical.interfaces import LogicalOperator
+from ray.types import ObjectRef
+
+if TYPE_CHECKING:
+    import pyarrow
+    import pyspark
+    import datasets
+
+
+class FromArrowRefs(LogicalOperator):
+    """Logical operator for `from_arrow_refs`."""
+
+    def __init__(
+        self,
+        tables: List[ObjectRef[Union["pyarrow.Table", bytes]]],
+        op_name: str = "FromArrowRefs",
+    ):
+        super().__init__(op_name, [])
+        self._tables = tables
+
+
+class FromArrow(FromArrowRefs):
+    """Logical operator for `from_arrow`."""
+
+    def __init__(
+        self, tables: List[Union["pyarrow.Table", bytes]], op_name: str = "FromArrow"
+    ):
+        super().__init__([ray.put(t) for t in tables], op_name)
+
+
+class FromSpark(FromArrowRefs):
+    """Logical operator for `from_spark`."""
+
+    def __init__(
+        self,
+        df: "pyspark.sql.DataFrame",
+        parallelism: Optional[int] = None,
+    ):
+        from raydp.spark.dataset import _save_spark_df_to_object_store
+
+        self._parallelism = parallelism
+
+        num_part = df.rdd.getNumPartitions()
+        if parallelism is not None:
+            if parallelism != num_part:
+                df = df.repartition(parallelism)
+        blocks, _ = _save_spark_df_to_object_store(df, False)
```

Comment:
Hmm doing things like this repartition and saving to the object store as part of the logical op construction feels like an anti-pattern, since logical ops should describe the "what" that we'd like to do (i.e. basically a logical config of the operation), not perform the actual execution to do it. We also don't want to be triggering (new) Ray work until we are actually executing the dataset, if possible.

Ideally we'd capture the Spark-on-Ray dataframe and parallelism in the logical operator, and do this actual execution to convert it to a `Dataset` with a `PhysicalOperator` governed by the executor, similar to what we do when we create a `Dataset` from read tasks. Not sure how this would work best with the existing `InputDataBuffer` and the like.

@ericl @c21 @jianoaix what do y'all think?

Reply:
Yeah, I had to break the typical pattern here because I was trying to mirror the existing implementation as closely as possible. Since `from_spark()` [boils down to](https://github.com/oap-project/raydp/blob/master/python/raydp/spark/dataset.py#L179-L180) generating blocks using the `_save_spark_df_to_object_store` method, then uses `from_arrow_refs(blocks)`, I tried to parallel this by having `FromSpark` extend the `FromArrowRefs` operator. I wasn't sure of another approach that would integrate with `InputDataBuffer`.

Reply:
I think you can make a subclass of InputDataBuffer that populates its inputs when it is started (on `start()`, perhaps).

Reply:
Oh wait, can't you put this into the `input_data_factory` arg already?

Reply:
Gotcha, I moved the above logic into `input_data_factory` (in [plan_from_spark_op.py](https://github.com/ray-project/ray/pull/32959/files#diff-1164caa96c63adfd1c422efbe9c87891dd7dd118e623fc2c3e6277db45c025e2R21-R27)), and now `FromSpark` no longer extends `FromArrowRefs`.